### PR TITLE
Resource and Data Source Support for Custom Redirects and Filters

### PR DESCRIPTION
**Pull request by unasra**
Added Terraform support for :
- Custom Redirects
- Application Filters
- Category Filters
- Content Categories ( Required for Category Filters )
- Threat Feeds

### CODE REVIEW

File: internal/service/fw/model_atcfw_application_filter.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+func (m *AtcfwApplicationFilterModel) Expand(ctx context.Context, diags *diag.Diagnostics) *fw.ApplicationFilter {
+	if m == nil {
+		return nil
+	}
+	to := &fw.ApplicationFilter{
+		Criteria:    flex.ExpandFrameworkListNestedBlock(ctx, m.Criteria, diags, ExpandAtcfwApplicationCriterion),
+		Description: flex.ExpandStringPointer(m.Description),
+		Name:        flex.ExpandStringPointer(m.Name),
+		Readonly:    flex.ExpandBoolPointer(m.Readonly),
```
Comment (mathewab) [inline internal/service/fw/model_atcfw_application_filter.go:116]: if it is "Computed/Read only" it shouldn't be Expanded.

File: internal/service/fw/api_application_filter_resource_test.go [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+}
+
+func testAccApplicationFiltersCriteria(name, criteriaName string) string {
+	return fmt.Sprintf(`
+resource "bloxone_td_application_filter" "test_criteria" {
+	name = %q
+	criteria  = [
+	{
+		name = %q
+	}
```
Comment (mathewab) [inline internal/service/fw/api_application_filter_resource_test.go:275]: can you also test for the other sub fields - category,subcategory

File: docs/data-sources/td_content_categories.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+# bloxone_td_content_categories (Data Source)
+
+
+
+## Example Usage
+
+```terraform
+# Get all Content Categories
+data "bloxone_td_content_categories" "all_content_categories" {}
```
[the remark this answers is not in the document]
Comment (unasra) [reply docs/data-sources/td_content_categories.md:17]: No Filters for this.

File: docs/data-sources/td_application_filters.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
+---
+# generated by https://github.com/hashicorp/terraform-plugin-docs
+page_title: "bloxone_td_application_filters Data Source - terraform-provider-bloxone"
+subcategory: "Threat Defense"
+description: |-
```
[the remark this answers is not in the document]
Comment (unasra) [reply docs/data-sources/td_application_filters.md:5]: Added

File: docs/resources/td_category_filter.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+Creates and Manages Category Filters.
+
+## Example Usage
+
+```terraform
+data "bloxone_td_content_categories" "test" {}
+
+resource "bloxone_td_category_filter" "example" {
+  name       = "example_category_filter"
+  categories = [data.bloxone_td_content_categories.test.results.0.category_name]
```
[the remark this answers is not in the document]
Comment (unasra) [reply docs/resources/td_category_filter.md]: Modified Tests and Examples to use string input

File: docs/resources/td_category_filter.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+  # Other optional fields
+  description = "Example of an Category Filter"
```
[the remark this answers is not in the document]
Comment (unasra) [reply docs/resources/td_category_filter.md]: Fixed

File: internal/provider/provider.go [hunk 7]
```diff
@@ -3,6 +3,7 @@ package provider
 import (
 	"context"
 	"fmt"
+	"github.com/infobloxopen/terraform-provider-bloxone/internal/service/redirect"
```
[the remark this answers is not in the document]
Comment (unasra) [reply internal/provider/provider.go]: Fixed

File: internal/service/fw/model_atcfw_application_filter.go (hunk 1 above)
Comment (unasra) [reply internal/service/fw/model_atcfw_application_filter.go]: Fixed

**Review by mathewab — APPROVED** [review]

**Pull request merged**